File: Python/Binary Trees/HeightOfBT.py

class Node:
    def __init__(self, data):
        self.data= data
        self.left = None
        self.right = None
    
def treeHeight(root):
    if root is None:
        return 0
    height = 0
    q = [(root, 0)]

    while q:
       node, level = q.pop(0)
       height = max(height, level)
       if node.left:
           q.append((node.left, level+1))
       if node.right:
           q.append((node.right, level+1))
    return height + 1

if __name__ == "__main__":
    root = Node(1)
    root.left = Node(2)
    root.right = Node(3)
    root.left.left = Node(4)
    root.left.right = Node(5)
    root.right.right = Node(6)
    root.left.right.left = Node(7)
    root.right.right.left = Node(8)
    root.right.right.left.left = Node(9)

    height = treeHeight(root)
    print('Height of BT: ', height)

File: C++/BInary Trees/HeightOfBT.cpp
#include <iostream>
#include <bits/stdc++.h>
// #include <queue>

using namespace std;

class Node
{
public:
    int data;
    Node *left;
    Node *right;

    Node(int val)
    {
        data = val;
        left = nullptr;
        right = nullptr;
    }
};

int Height(Node *root)
{
    if (root == nullptr)
        return 0;

    queue<pair<Node *, int>> q;
    int height = 0;
    q.push({root, 0});

    while (!q.empty())
    {
        Node *node = q.front().first;
        int level = q.front().second;
        q.pop();

        height = max(height, level);

        if (node->left)
        {
            q.push({node->left, level + 1});
        }

        if (node->right)
        {
            q.push({node->right, level + 1});
        }
    }
    return height + 1;
}

void deleteTree(Node *root)
{
    if (root == nullptr)
        return;

    deleteTree(root->left);
    deleteTree(root->right);
    delete root; // Delete the current node AFTER deleting children
}

int main()
{
    Node *root = new Node(1);
    root->left = new Node(2);
    root->right = new Node(3);
    root->left->left = new Node(4);
    root->left->right = new Node(5);
    root->right->right = new Node(6);
    root->left->right->left = new Node(7);
    root->right->right->left = new Node(8);
    root->right->right->left->left = new Node(9);

    int ans = Height(root);
    cout << "Height of BT is: " << ans << endl;

    // ---------------------------

    deleteTree(root);

    return 0;
}